### ISSUE TITLE
Strange fonts height in GNU/Linux

### ISSUE THREAD

**Issue by s8321414** [adobe-fonts/source-han-sans#114]
My environment: Chakra GNU/Linux using KDE
Using Source Han Sans OTC 1.004 (Source Han Sans TC variant)
It's screenshot:
![16](https://cloud.githubusercontent.com/assets/9457283/8325179/dbcef62e-1a89-11e5-9c2c-dc40f6cc42fe.png)
![17](https://cloud.githubusercontent.com/assets/9457283/8325180/dbd80002-1a89-11e5-8b24-bb90a38e4b95.png)
But for 1.002 (Source Han Sans TC):
![18](https://cloud.githubusercontent.com/assets/9457283/8325209/01b18a3c-1a8a-11e5-83a6-bf9821bff796.png)
![19](https://cloud.githubusercontent.com/assets/9457283/8325208/01af910a-1a8a-11e5-8a77-87555b875ca8.png)
I think something is wrong between 1.002 and 1.003, can you figure it out and fix it?

**Comment by zerng07** [comment]
Is that a problem of the font, or the problem of the rendering engine? 

It should be cleared out first. Needs some investigation though.

**Comment by s8321414** [comment]
@zerng07 I just downgrade otc-source-han-sans from 1.004 to 1.002. And all fine here. I don't do any other changes on my PC, so it should be font issue.
Qt and Gtk toolkit apps both have the same issue, so it should not be toolkit issue.

**Comment by s8321414** [comment]
@felixonmars How do you think about this issue?

**Comment by kenlunde** [comment]
Font vertical metrics requirements—or, perhaps more accurately, assumptions—are clearly different for UIs in that a much tighter line spacing is necessary or expected.

Ignoring the harmonization across all seven weights that was done in Version 1.003 and carried forward in Version 1.004, the Version 1.002 fonts effectively had two sets of vertical metrics, and it seems that which set was used depends on the client. One set was those specified by _OS/2.sTypoAscender_ and _OS/2.sTypoDescender_, which specified a 1000-unit em-box by using values of 880 and -120, respectively. These same values were also reflected in the _hhea.Ascender_ (880) and _hhea.Descender_ (120) fields (this is default AFDKO _makeotf_ behavior). The other set was those specified by _OS/2.usWinAscent_ and _OS/2.usWinDescent_, which was calculated by first removing excessively tall vertical-use glyphs from the calculation (but not yet harmonized across all weights).

For the Version 1.003 fonts, the _hhea.Ascender_ and _hhea.Descender_ values were simply overridden to match the _OS/2.usWinAscent_ and _OS/2.usWinDescent_ values (and harmonized across all weights). This is the same practice that is employed for the other Source families (Source Sans Pro, Source Serif Pro, and Source Code Pro), and results in much more consistent cross-platform vertical metrics for general use.

UI use is a whole new beast, and has very unique requirements, which is why most fonts that are intended for UI use use special values that result in the desired effect, or the layout engine of the UI is intelligent enough to do the right thing. In my opinion, The Right Thing™ for UIs is to use the _OS/2.sTypoAscender_ and _OS/2.sTypoDescender_ values, ignore the _OS/2.sTypoLineGap_ value, and pick a very modest values, say 200, for the line gap.

In summary, a single set of vertical metrics settings cannot possibly satisfy all possible usage scenarios, especially those for UIs that require very tight inter-line spacing, so this is either a *DIY* (Do It Yourself) project, or use a derivative project, such as @ShikiSuen's [OriginHanSansUI](https://github.com/ShikiSuen/OriginHanSansUI), for your purposes. Also, because general use far outweighs UI use, I am flagging this *As Designed*.

**Comment by ShikiSuen** [comment]
I have to explain something:

1. Due to bad access stability between GitHub and myself while I am in PRC, I migrated my OriginHanSansUI into https://originhansansui.codeplex.com/ , which reflects all changes @kenlunde asked me regarding 'Source' trademark. This change will not apply onto its GitHub repo until this late August.
![image](https://cloud.githubusercontent.com/assets/3164826/8331643/fd806c4e-1abb-11e5-8312-f122fcee5344.png)

2. Origin Han Sans UI is absolutely not a choice for GNU Linux's GUI fonts. You have to fork something for your own. I initially let Origin Han Sans use (HHEA: Asc=1000, Desc=-234) in all 7 weights, but that would cause problem like https://github.com/adobe-fonts/source-han-sans/issues/102 . **Thus, in the most-recent update "Aiza Nagi" (commited just now at its CodePlex website), I apply SHS's factorial (HHEA: Asc=1160, Desc=-320) metrics onto Origin Han Sans UI's Heavy and Bold weight. This MUST cause troubles like @s8321414 encountered & shown in his screenshots.**

P.S.: I will appreciate it if @kenlunde could tell me the minimum compatible metric sets of HHEA Ascender & Descender for SHS only.

**Comment by ShikiSuen** [comment]
One more thing: I don't know whether GNU Linux could handle the lineGap metrics, but I at least know OS X couldn't handle it correctly in most cases. Just want to let myself not be misunderstood.

**Comment by zerng07** [comment]
No problems with GNOME desktop with 1.004R. KDE-only problem doubted.

**Comment by zerng07** [comment]
Though the line gap does change, but there is no problems with rendering as an UI font with GNOME.

**Comment by ShikiSuen** [comment]
Thanks, then that might be KDE's problem.





On Thu, Jun 25, 2015 at 7:56 AM -0700, "zerng07" <notifications@github.com> wrote:
Though the line gap does change, but there is no problems with rendering as an UI font with GNOME.

---

**Comment by zerng07** [comment]
By the way, the new line gap in 1.004R looks better than 1.002R in gedit (a text editor in GNOME). The line gap in 1.002R is too close to read in gedit, and the new line gap works very well!

**Comment by ShikiSuen** [comment]
That's not line-gap-determined enhancement, please read Adobe official update description thoroughly.

從 Outlook 傳送




On Thu, Jun 25, 2015 at 8:03 AM -0700, "zerng07" <notifications@github.com> wrote:
By the way, the new line gap in 1.004R looks better than 1.002R in gedit (a text editor in GNOME). The line gap in 1.002R is too close to read in gedit, and the new line gap works very well!

---

**Comment by zerng07** [comment]
I mean the displaying result as referred to the line gap.

**Comment by ShikiSuen** [comment]
@zerng07 You said:
> By the way, the new line gap in 1.004R looks better than 1.002R in gedit (a text editor in GNOME). The line gap in 1.002R is too close to read in gedit, and the new line gap works very well!

Based on what Adobe claimed among their official SHS update histories:
- 1.002R use linegap=0 in both HHEA and OS/2 tables;
- 1.003R change Ascender & Descender in **HHEA** table from [880, -120] to [1160, -320];
- 1.004R only fixed some glyph problems regarding Japanese version of SHS 1.003R.

Even though you are talking about the display result, you still should use "the gap between lines" (not the terminology "lineGap") to avoid misleading your possible readers.

You make comparison between 1.002R and 1.004 (regarding their display effect in gedit), but their lineGap metrics are all zero. You should realize that it's because of the update made in 1.003R (as what I mentioned above).

**Comment by ShikiSuen** [comment]
Key translation regarding my last reply:

1. 是說 LineGap 一詞乃術語。
2. 您說「思源黑體 1.004R 相對 1.002R 而言，在 gedit 當中的行間距更舒服」，這是因為 1.003R 在 HHEA 表當中啟用了全新的 Ascender & Descender 數值所致（注意我在這裡只說了 HHEA 表，1.003R 並沒有針對 OS/2 表做出任何修改）。
3. 1.004R 僅包含了對日文版思源黑體的縱排假名 Glyph 修正。

**Comment by ShikiSuen** [comment]
去睡覺了，希望您能夠明白我剛才所解釋的。
(Go sleep, hope that you could understand my explanations above.)

**Comment by zerng07** [comment]
Hi, I know the difference between 1.003R and 1.004R, so there is no difference here. I also follow the issues here, so I know what had happened.

And sorry for the terms misleading here, I apologize for that. I was just trying to convey that the result is even better with the new metrics in GNOME. And that's all.

**Comment by ShikiSuen** [comment]
to @zerng07 >

The fact you claimed could be a good reference to Ken Lunde, so that he could still insist on thinking of "It's still because of reading wrong metrics from the specific font".

It's time to send a bug report to KDE team to let them take a look at this thread.

**Comment by ShikiSuen** [comment]
(I modified my last reply, please check it at GitHub website but not your mail.)

**Comment by kenlunde** [comment]
One of the advantages of Source Han Sans is that it pushes the limits in a variety of ways, in terms of the number of glyphs (at the architectural limit), broader coverage of Unicode, broader coverage of vertical glyphs, and so on. The result is that poor assumptions are being exposed. Conventional fonts may also expose such poor assumptions, but in more subtle ways, meaning that they are more difficult to notice.

The excessively tall vertical glyphs in Source Han Sans, in particular, are very useful in exposing environments that use the FontBBox (as expressed by _head.xMin_, _head.yMin_, _head.xMax_, and _head.yMax_) for line-layout purposes, which is a huge no-no. These are U+2E3A, U+3E3B, U+3031, and U+3032.

The Latin glyphs for Vietnamese, which use multiple accents, are primarily the reason for the ascender values.

**Comment by s8321414** [comment]
hmmm, so it shoule be KDE or Qt specific bug, should report it to their bugtracker.

**Comment by ryuanlu** [comment]
This is my screenshot. I have no idea if it's the same problem or not.

fontconfig is configured to use Liberation Sans for Latin and Source Han Sans for CJK.

![screenshot from 2015-07-18 19 06 32](https://cloud.githubusercontent.com/assets/12210652/8761358/2f4fbe88-2d80-11e5-948c-888ed48533b0.png)

**Comment by ShikiSuen** [comment]
@ryuanlu If you are using GNOME, you met another issue.
If you are using KDE, probably you met the same issue.

**Comment by zerng07** [comment]
@ryuanlu What version of Source Han Sans you use? And how do you set the fontconfig configuration.

I don't think this is much related to Source Han Sans.

**Comment by ryuanlu** [comment]
@ShikiSuen Yes, I'm using GNOME 3.14 (Debian Jessie)

@zerng07  1.004R

/etc/fonts/local.conf https://gist.github.com/ryuanlu/2c90dfdf554215f6257c

Audacious 3.5 by Debian, not Qt build.
Not only audacious, but also gnome-shell (clutter), steam.

Maybe it's related to pango ?

![screenshot from 2015-07-18 20 31 17](https://cloud.githubusercontent.com/assets/12210652/8761600/02b84942-2d8c-11e5-81d5-f0cce3d38bac.png)

**Comment by zerng07** [comment]
It seems that you don't have the font height cutting problem in the second screenshot.

I don't see any problem in the second screenshot, could you just point out?
What have you done between these two screenshots?

Note: You have specified Liberation Sans before Source Han Sans for Sans, and please remember that Liberation Sans doesn't share the same metrics with Source Han Sans.

**Comment by ryuanlu** [comment]
In the first screenshot, I set font size to 16 to emphasize the difference.
In the second screentshot, font size is reset to 10.

In some cases, CJK glyphs are a bit lower than Latin ones.

![screenshot from 2015-07-18 20 31 17](https://cloud.githubusercontent.com/assets/12210652/8761993/6f0dc06a-2d9e-11e5-977b-03b31eda419c.png)

**Comment by zerng07** [comment]
Yeah, remember that according to your fontconfig configuration your Latin ones are set to Liberation Sans. 

The problem might be caused by that the app is using the matrix from Liberation Sans to display the box, and Source Han Sans is using another different matrix.

I suggest you using Source Han Sans directly and removing all the bindings to see if it fixes your problem.

**Comment by ryuanlu** [comment]
Ok, just curious why most cases are not effected.

**Comment by zerng07** [comment]
What cases are not infected?

Some programs such as the ones written in Qt will only pick the most preferred font in your fontconfig conf according to its internal rule, and neglecting those binding sections and fallback sections. They only use one font for UI, so you won't see the CJK glyphs a bit lower than Latin ones.

**Comment by ryuanlu** [comment]
VLC works with my fontconfig conf (Liberation Sans + SHS TW)

![untitled](https://cloud.githubusercontent.com/assets/12210652/8762564/616cdf70-2daf-11e5-8bac-4f319a7ca08b.png)
![screenshot from 2015-07-19 00 39 42](https://cloud.githubusercontent.com/assets/12210652/8762601/9aa7cfc4-2db0-11e5-83fa-5fd2b7f727b5.png)

**Comment by zerng07** [comment]
VLC is based on Qt, and Qt does not respect fontconfig. What you see here is using one font (should be Source Han Sans in your case) for UI.

**Comment by zerng07** [comment]
What do you mean "VLC works"? I guess you have the same problem with "7000 個區塊".

Qt might fix something relating to fontconfig that it may result like GTK does. Anyway, using Source Han Sans for sans-serif directly and removing the binding section might save you from this problem and you can try.

The last but not the least, I still don't think there is much to do with Source Han Sans in this case.

**Comment by ShikiSuen** [comment]
@ryuanlu This is not fine. At least I could see terrible issue of misaligned horizontal-axis among fallbacked glyphs and unfallbacked glyphs. I met this issue by using both Gnome / Cinnamon with default western font "Source Sans Pro" and its fallbacked font "Source Han Sans TC", this issue still exists. The font layout engine in Linux at this moment is extremely ridiculous since Source Han Sans and Source Sans Pro share the same metrics (refer to what @kenlunde claimed in the release notes of SHS 1.004).

**Comment by ShikiSuen** [comment]
@kenlunde Would it be a good idea to suggest Fedora development team to use Source Sans Pro as their default UI font? This could probably make them realize the issue of their layout engine regarding misaligned horizontal-axis, while they are currently using Source Han Sans regional subsets for Chinese UI fonts as default in Fedora (I know nothing about Red Hat Enterprise Builds).

**Comment by ShikiSuen** [comment]
Also, the screenshot attached in the 1st post also proved that KDE also has the issue of their layout engine regarding misaligned horizontal-axis. (See the upper-right date and time elements, they are not aligned horizontally. Note that this is definitely not a matter of font size.)

**Comment by s8321414** [comment]
@ShikiSuen no... it's KDE4

**Comment by ShikiSuen** [comment]
@s8321414 Would it be convenient for you to test KDE Plasma 5? (It may not necessary if KDE use the same font rendering solution (Pango) as what GNOME 3 used) Thanks.

**Comment by s8321414** [comment]
@ShikiSuen OK I may do it later.

**Comment by s8321414** [comment]
1.002 on KDE Plasma 5:
![screenshot_20160323_175041](https://cloud.githubusercontent.com/assets/9457283/13983142/e11b505c-f128-11e5-81f0-4fad0841d6de.png)
1.004 on KDE Plasma 5:
![screenshot_20160323_174707](https://cloud.githubusercontent.com/assets/9457283/13983158/02a05a6a-f129-11e5-87ce-ff75f68b8505.png)

**Comment by s8321414** [comment]
hmmm... So it should be freetype issue.

**Comment by kenlunde** [comment]
If it is an alignment issue, I don't see how it can be a FreeType issue, but rather it's more likely to be a layout engine issue in the app or in a library.

**Comment by ShikiSuen** [comment]
@kenlunde I prefer to think that it's an issue of a shared library since I saw such issue in different apps and shell solutions (incl. Cinnamon desktop).

**Comment by ShikiSuen** [comment]
@s8321414 Note that the align issue may not be able to show up since you let Source Han Sans be default western UI font in its highest priority. (i.e. The current western font is Source Han Sans, no kanji fallback triggered)

**Comment by ShikiSuen** [comment]
@s8321414 Could you please tell me that which version of Source Han Sans are you using in your screenshot of KDE Plasma 5? Thanks.

**Comment by s8321414** [comment]
@ShikiSuen Both are on Plasma 5.5.5 in latest Chakra GNU/Linux, I have the same issue on BBSFox in Firefox on Windows...

**Comment by ShikiSuen** [comment]
@s8321414 Thanks for your screenshot. After comparison to each other, I am still wondering whether you have initial issue in KDE Plasma 5.5.5 with SHS 1.004. Looks like some layout method towards date & time gadgets are changed in KDE 5.

**Comment by s8321414** [comment]
I don't know, but it's really weird...

**Comment by ShikiSuen** [comment]
Ubuntu 16.04 LTS adopts Noto Sans CJK as default CJK UI font. In such case, any font layout error shall be appealed to the font layout engine used.

**Comment by kenlunde** [comment]
@ShikiSuen: Thank you. I appreciate the heads up.

**Comment by s8321414** [comment]
Well, this still happened in 5.14 series... I have no idea if SHS 2.000 resolves this issue.

**Comment by s8321414** [comment]
@ShikiSuen BTW, where is your Origin Han Sans UI repo now?

**Comment by kenlunde** [comment]
@s8321414 As this is clearly not a font issue, the Version 2.000 fonts are unlikely to resolve this.

**Comment by ShikiSuen** [comment]
@s8321414

This project is temporarily dead.

In these years I gave up waiting for SHS (also due to my satisfaction with PingFang family) and was focusing on my school studies. Now I graduated as a Bachelor of Music and is busy studying Nihongo to fit my needs of further-studying music in Japan in the next two years.

If you have interests in it, you can copy the changed parameters in the `hhea` table and apply it into Source Han Sans 2.0 builds. Remember, you have to change the font name before distributing your modified fonts (I forgot the reason but I guess Dr. Lunde can explain).
